helpers/utils.py

```diff
@@ -75,36 +113,401 @@ def parse_conf_overrides(conf_args: List) -> Dict:
     return conf_overrides
 
 def reconstruct_submit_args(args: List, conf: Dict) -> List:
+    """Adding back possibly overridden config properties to list of other spark-submit arguments
+
+    Args:
+        args: regular args passed to spark-submit
+        conf: dictionary of all additional config (possible after overrides) to be passed to spark-submit.
+    """
     submit_args = args
     conf_arg = ''
     for k in conf.keys():
         conf_arg += f' --conf {k}={conf[k].strip()}'
     submit_args = [conf_arg] + submit_args
     return submit_args
 
-def autodetect_kubernetes_master(conf: Dict) -> str:
+def get_kube_config() -> str:
+    """Returns default kubeconfig to use if not explicitly provided."""
     USER_HOME_DIR = pwd.getpwuid(os.getuid())[USER_HOME_DIR_ENT_IDX]
     DEFAULT_KUBECONFIG = f'{USER_HOME_DIR}/.kube/config'
     kubeconfig = os.environ.get('KUBECONFIG') or DEFAULT_KUBECONFIG
+    return kubeconfig
+
+def get_kubectl_cmd() -> str:
+    """Returns the kubectl binary location within the snap, used to build k8s commands."""
     kubectl_cmd = '{}/kubectl'.format(os.environ['SNAP'])
+    return kubectl_cmd
+
+def autodetect_kubernetes_master(conf: Dict) -> str:
+    """Return a kubernetes master for use with spark-submit in case not provided.
+
+    Args:
+        config: dictionary of all config available to spark-submit.
+    """
+    kubeconfig = get_kube_config()
+    namespace = conf.get('spark.kubernetes.namespace')
     context = conf.get('spark.kubernetes.context')
-    if context:
-        master_kubectl_cmd = f"{kubectl_cmd} --kubeconfig {kubeconfig} --context {context} config view --minify -o jsonpath=\"{{.clusters[0]['cluster.server']}}\""
-    else:
-        master_kubectl_cmd = f"{kubectl_cmd} --kubeconfig {kubeconfig} config view --minify -o jsonpath=\"{{.clusters[0]['cluster.server']}}\""
-    try:
-       default_master = subprocess.check_output(master_kubectl_cmd, shell=True)
-       k8s_master_uri = 'k8s://' + default_master.decode("utf-8")
-    except subprocess.CalledProcessError as e:
-       logging.error(f'Cannot determine default kubernetes control plane url. Probably bad or missing kubeconfig file {kubeconfig}')
-       logging.warning(f'Please set the KUBECONFIG environment variable to point to the kubeconfig. Or fix the default kubeconfig {kubeconfig}')
-       logging.error(e.output)
-       sys.exit(EXIT_CODE_BAD_KUBECONFIG)
-    return k8s_master_uri
+    primary_sa = retrieve_primary_service_account_details(namespace, kubeconfig, context)
+    primary_namespace = primary_sa.get('spark.kubernetes.namespace') or 'default'
+    kubectl_cmd = build_kubectl_cmd(kubeconfig, primary_namespace, context)
+    master_cmd = f"{kubectl_cmd} config view --minify -o jsonpath=\"{{.clusters[0]['cluster.server']}}\""
+    default_master = execute_kubectl_cmd(master_cmd, EXIT_CODE_BAD_KUBECONFIG)
+    return f'k8s://{default_master}'
 
 def UmaskNamedTemporaryFile(*args, **kargs):
```

Comment:
I know it does not pertain this PR...but I suppose we should go for the lower case, snake convention here, being a python function

Reply:
noted.